Cache file upload original names

MartinRiese opened the pull request:
## Product Description

When deleting a repeat group we need to rebuild all the following groups because the IDs are reassigned by formplayer. Unfortunately FP does not keep track of the original filename automatically and only sends the uuid based generated filename for the upload in the delete-repeat response. To be able to still display the original filename we store it in a map with the generated name as the key. That way we can figure out what it is when we rebuild that part of the UI.

## Technical Summary

https://dimagi.atlassian.net/browse/USH-5078

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

tested locally and on staging

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

### Automated test coverage

corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/form_ui_spec.js

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

MartinRiese committed:
Cache file upload original names

When deleting a repeat group we need to rebuild all the following groups
because the IDs are reassigned by formplayer. Unfortunatly FP does not
keep track of the original filename automatically and only sends the
uuid based generated filename for the upload in the delete-repeat
response. To be able to still display the original filename we store it
in a map with the generated name as the key. That way we can figure out
what it is when we rebuild that part of the UI.

Robert-Costello reviewed (approved):
This looks good to me. Not blocking, but is it possible to add a test/assert that checks the filename is cached and/or reassigned after a repeat is deleted?

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/entries.js:
```diff
@@ -948,17 +949,26 @@ hqDefine("cloudcare/js/form_entry/entries", [
+            const fixedNewValue = newValue.replace(constants.FILE_PREFIX, "");
             self.fileNameDisplay(fixedNewValue);
             self.answer(fixedNewValue);
+        } else if (cachedFilename) {
```
Do you know what the code above is doing? What situation is `newValue` an empty string vs `constants.NO_ANSWER` (null)?

```
        if (newValue === "" && self.question.filename) {
            self.question.hasAnswered = true;
            self.fileNameDisplay(self.question.filename());
   ```
It looks like Minha added this previously [here](https://github.com/dimagi/commcare-hq/pull/35041/files) to solve a similar issue with the file question being rebuilt for question tiles. I'm not understanding why that doesn't solve the same issue here.

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/entries.js:
```diff
@@ -938,7 +938,8 @@ hqDefine("cloudcare/js/form_entry/entries", [
     FileEntry.prototype.onPreProcess = function (newValue) {
-        var self = this;
+        const self = this;
+        const cachedFilename = self.question.form().fileNameCache[self.answer()];
```
what's the distinction between `self.answer()` and `self.question.answer()`?

MartinRiese replied:
There is none. self.answer just references to self. question.answer

MartinRiese committed:
Merge branch 'master' into riese/filename_broadcast

MartinRiese committed:
Add condition I missed in last commitAdd condition I missed in last
commitAdd condition I missed in last commitAdd condition I missed in
last commitAdd condition I missed in last commitAdd condition I missed
in last commitAdd condition I missed in last commitAdd condition I
missed in last commit
